### PR TITLE
RFC: incremental compilation

### PR DESCRIPTION
**Pull request by ChristopherBiscardi**
Toast should be thought of as a compiler and we should adopt modern compiler practices that make sense like incremental compilation from the start. It will be very hard to add this "after the fact".

### CODE REVIEW

File: rfcs/0000-incremental-compilation.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
+- Feature Name: incremental-compilation
+- Start Date: 2020-03-15
+- RFC PR:
+  [christopherbiscardi/toast#0000](https://github.com/christopherbiscardi/toast/pull/0000)
```
Comment (ChristopherBiscardi) [inline rfcs/0000-incremental-compilation.md]:
```suggestion
  [christopherbiscardi/toast#1](https://github.com/ChristopherBiscardi/toast/pull/1)
```

File: rfcs/0000-incremental-compilation.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
+- Feature Name: incremental-compilation
+- Start Date: 2020-03-15
+- RFC PR:
+  [christopherbiscardi/toast#1](https://github.com/ChristopherBiscardi/toast/pull/1)
```
Comment (ChristopherBiscardi) [inline rfcs/0000-incremental-compilation.md]:
```suggestion
  [christopherbiscardi/toast#0001](https://github.com/ChristopherBiscardi/toast/pull/1)
```

**Comment by tomByrer** [comment]
Interesting!

Ideas for flagging what files to render (assuming others would be left out).

* last x days/hours
* files changed via `git`
* manual

From the times I played with GatsbyJS, it seems images really slow things down.  Just text blogs tend to re-render fast.

**Comment by ChristopherBiscardi** [comment]
> Ideas for flagging what files to render (assuming others would be left out).

I think these are all optimizations. The first step is to implement the caching such that (for example) when a file gets read, it gets hashed, if that hash matches it doesn't trigger any further processing. So basically implementing that hashing for the results of computations through the entire pipeline. 

> From the times I played with GatsbyJS, it seems images really slow things down. Just text blogs tend to re-render fast.

yeah images tend to slow down gatsby builds and the way to fix that is to not handle images or to split up the image processing and parallelize it (which is only available in their Cloud product). There's also the issues with sharp installs, etc.

**Comment by ChristopherBiscardi** [comment]
Merging this as the Rust prototype has a working core for it. It doesn't look like the description in this PR (ex: no merkle trees) but enables the feature as a whole

**Pull request merged**